returnn/tf/layers/basic.py

```diff
@@ -1176,6 +1176,143 @@ def transform_config_dict(cls, d, network, get_layer):
       d["position"] = get_layer(d["position"])
 
 
+class SliceNdLayer2(_ConcatInputLayer):
+  """
+  This takes out a slice-range from some axis,
+  e.g. ``x[start:start + size]``.
+  This layers allows a different start slice point for each batch,
+  in contrast to :class:`SliceLayer`, and the start is variable.
+  See also :class:`GatherNdLayer`.
+  :class:`PrefixInTimeLayer` can recover the original shape (by zero-padding).
+
+  Gathers slices on a specified axis from the input layer using indices from a ``position`` layer.
+  If the input is a layer of the shape ``[B,T1,T2,F1]``, and start of shape ``[B,T1]``, and size is S,
+  this will yield output of the shape ``[B,T1,S,F1]`` where
+
+  ``output[b,t1,0:S,f1] = input[b,position[b,t1],T2:T2+S,f1]``
+
+  (if ``D`` is the axis to gather from).
+  In general, all shared axes of the input and the positions will be considered as batch-axes.
+
+  The ``position`` argument can also be an ``int``.
+  In this case, this simply gives ``input[position]`` one the specified ``axis``.
+
+  It's basically a wrapper around ``tf.gather``.
+  It provides the same functionality as the deprecated ``GatherNdLayer``, but is more generic.
+  See also :class:`GatherNdLayer`.
+  """
+  layer_class = "slice_nd2"
+  recurrent = True
+
+  def __init__(self, start, size, min_size=None, **kwargs):
+    """
+    We expect the input to have at least one more axis than start and the rest of the axis in front are the same.
+    In case the input has no extra time axis compared to start, we assume slice_nd is pulled out of a rec layer
+    but the input has stood the same.
+
+    :other LayerBase input_data:  shape [B,T0,..,Tn,D] or [B,T0,..,Tn-1,D]
+    :param LayerBase start:  shape [B,T0,..,Tn-1]
+    :other LayerBase output: shape [B,T0,..,Tn',D]
+    :param int|None size: if None, it uses the max possible size, and it becomes a dynamic axis
+    :param int|None min_size: if size is None, but we want to have a min-size, set this
+    """
+    super(SliceNdLayer2, self).__init__(**kwargs)
+    from returnn.tf.util.basic import slice_nd2, DimensionTag
+    assert start.output.have_batch_axis() and self.input_data.have_batch_axis()
+    self.start = start
+
+    input_data = self.input_data.copy_as_batch_major()
+    start = start.output.copy_as_batch_major()
+
+    # make sure axis of start are in input
+    is_equal_opts = dict(ignore_feature_dim=True, allow_same_spatial_dim=True, broadcast_matches=True)
+    for start_axis in range(start.batch_ndim):
+      assert input_data.get_dim_tag(start_axis).is_equal(start.get_dim_tag(start_axis), **is_equal_opts)
+
+    # Handle the case when layer is pulled out of rec loop but the input hasn't change
+    if self.optimized_out_of_loop_and_unchanged_input(input_data, start):
```

Comment:
There should be nothing about "optimized-out-of-loop" in this layer.
It should not matter at all whether this is in a rec layer.
The layer just gets some input and should do the correct thing.

Reply:
Well, I explained above the discrepancy. 

There exists both the possibility that the input layer to `slice_nd` gets pulled out or that it comes from `base:network` and stays the same, when `slice_nd` gets pulled out.

I can rename the function to something like: `input_needs_tiling`